napalm_junos/junos.py

```diff
@@ -464,33 +475,95 @@ def _parse_value(value):
+            also when the device is capable to return the routing
+            instance name at the BGP neighbor level.
+            '''
+            for bgp_neighbor in neighbor_data:
+                peer_ip = napalm_base.helpers.ip(bgp_neighbor[0].split('+')[0])
+                neighbor_details = deepcopy(default_neighbor_details)
+                neighbor_details.update(
+                    {elem[0]: elem[1] for elem in bgp_neighbor[1] if elem[1] is not None}
+                )
+                if not instance:
+                    # not instance, means newer Junos version,
+                    # as we request everything in a single request
+                    peer_fwd_rti = neighbor_details.pop('peer_fwd_rti')
+                    instance = peer_fwd_rti
+                else:
+                    # instance is explicitly requests,
+                    # thus it's an old Junos, so we retrieve the BGP neighbors
+                    # under a certain routing instance
+                    peer_fwd_rti = neighbor_details.pop('peer_fwd_rti', '')
+                instance_name = 'global' if instance == 'master' else instance
+                if instance_name not in bgp_neighbor_data:
+                    bgp_neighbor_data[instance_name] = {}
                 if 'router_id' not in bgp_neighbor_data[instance_name]:
                     # we only need to set this once
                     bgp_neighbor_data[instance_name]['router_id'] = \
-                        py23_compat.text_type(neighbor_data['local_id'])
+                        py23_compat.text_type(neighbor_details['local_id'])
                 peer = {
                     key: self._parse_value(value)
-                    for key, value in neighbor_data.items()
+                    for key, value in neighbor_details.items()
                     if key in keys
                 }
-                peer['address_family'] = self._parse_route_stats(neighbor_data)
+                peer['address_family'] = self._parse_route_stats(neighbor_details)
+                if 'peers' not in bgp_neighbor_data[instance_name]:
+                    bgp_neighbor_data[instance_name]['peers'] = {}
                 bgp_neighbor_data[instance_name]['peers'][peer_ip] = peer
-            for neighbor, uptime in uptime_table.get(instance=instance).items():
-                bgp_neighbor_data[instance_name]['peers'][neighbor]['uptime'] = uptime[0][1]
+                if not uptime_table_items:
+                    uptime_table_items = _get_uptime_table(instance)
+                for neighbor, uptime in uptime_table_items:
+                    if neighbor not in bgp_neighbor_data[instance_name]['peers']:
+                        bgp_neighbor_data[instance_name]['peers'][neighbor] = {}
+                    bgp_neighbor_data[instance_name]['peers'][neighbor]['uptime'] = uptime[0][1]
+
+        old_junos = napalm_base.helpers.convert(
+            int, self.device.facts.get('version', '0.0').split('.')[0], 0) < 13
```

Comment:
I would be inclined to put the old_junos variable and a brief comment at the top of the function (it probably makes it easier to understand). We might even want to set it as an attribute of the object that we set at open (since we use it in two methods).